### ISSUE TITLE
too slow?

### ISSUE THREAD

**Issue by manageseverin** [nsarrazin/serge#65]
I'm using it under Windwos 11 with alpaca 7B
Ok, it's great overall, but I have a native cpp version (chat.exe) and it's running 2 times faster than your docker version. 
Also, how to use the API ? I saw in docker something like 127.0.0.1:35272 - "GET /chat/5fe89704-c7ca-4a67-9ec2-f267689b0ffe/question?prompt=No%2C+it%27s+actually+14 HTTP/1.1" 200 OK
But where to look for proper API documentation ?

**Comment by mokahless** [comment]
Ubuntu 22. using 30B

While a significant amount of this issue appears to be CPU processing, I am watching resource usage and noticing every time it finishes an answer, it unloads the model from ram, and so for every question I ask it, it has to read the entire thing from disk into ram again. For my system, this wastes about 15 seconds every question. This portion seems to scale with the threads you give it, too. The same attempt with a quarter of the threads took ~25 seconds to move from disk to ram.

Seems to me this could be sped up by keeping the model in ram? Is that possible or is there something more complex going on that can't be improved upon?

**Comment by y12studio** [comment]
@manageseverin 

Try this docs URL. Serge - Swagger UI http://localhost:8008/api/docs

**Comment by manageseverin** [comment]
On alpaca 7B model, it takes about a minute for the answer to start to appear. Win 11, Ryzen 5600G/16GB RAM
I thought that it depends on the history size it need to feed back, but no - even for the first time (when there is no history) it also takes the same time.

**Comment by magicmars35** [comment]
@nsarrazin Hi Nathan, do you think it's possible to load the model in RAM, and keep it in ram, as long as the user querries are submitted ?

**Comment by nsarrazin** [comment]
Yes it's on the list!

**Comment by BrunoIsaac27** [comment]
hey, it's really slow! amd epyc 16gb ram, and a lot of delay. more than a minute to load the model to ram. (SSD)

**Comment by BrunoIsaac27** [comment]
wtf? I have a powerful server! how heavy is this?
![image](https://user-images.githubusercontent.com/66482679/228143475-9863f5b1-82b6-47cd-9e3e-16987aba1257.png)

**Comment by mausch** [comment]
It seems it doesn't use more than 4 CPUs. Maybe the `-t` flag needs to be applied? https://github.com/antimatter15/alpaca.cpp/issues/154#issuecomment-1483629705

**Comment by futurepr0n** [comment]
![image](https://user-images.githubusercontent.com/478986/228281215-4d997691-65a9-469b-89eb-7ddabe544d88.png)
I am also fairly beefy specs - I find it slow to run but I am guessing its due to the issue pointed out offloading the model. If we could keep it persistent it would work best, but I could see the reason for not wanting that maybe that its tricky to manage multiple sessions if you implement it right now? I don't know - maybe we could have a superuser option if we have the available ram to keep it loaded full time?

**Comment by BrunoIsaac27** [comment]
the problem is not serge, it's in llama.cpp. Something doesn't work well with docker, I saw 'npx dalai serve' run and the model responded in 3-5 seconds.. with docker here on my server, it took between 18-60min to initialize and load to ram, and 8 minutes for the model to finish its relatively small response.

**Comment by Mattssn** [comment]
![cIEGaZy](https://user-images.githubusercontent.com/54347017/228514023-f158a426-ee15-48b7-b658-e2d777b3f635.png)

I am playing with the 30B Model, and I am having the same thing, I am running this on Docker on a pretty beefy box, but getting pretty slow response times. Something to maybe have a variable to keep the model alive for 10 minutes then shut it down for inactivity, tbh, my server really never passes like 14gb of memory usage, an option to keep it loaded as long as the docker container is running might be cool too, I know you said you are working on it, just wanted to give my feedback :)

**Comment by voarsh2** [comment]
Quite unusable to constantly need to load/unload 4GB's in RAM. Not everyone is on SSD either, so then you have to contend with waiting on disks as well to load in RAM for every chat submission.

**Comment by BrunoIsaac27** [comment]
I installed it outside of docker and did not get different results from those already mentioned above. What is it? Because I've seen some people running alpaca 7B and it loads and responds in seconds, and on my machine which is a powerful computer for servers, even the 7B is extremely slow, even consuming 6 cores (I upgraded from intel xeon to amd epyc which reduced the answer to 8 minutes and the load to 18 minutes)

**Comment by johncadengo** [comment]
@voarsh2 you mentioned the default threads are 4. Where is this located? Is there a way to change it?

**Comment by voarsh2** [comment]
> @voarsh2 you mentioned the default threads are 4. Where is this located? Is there a way to change it?

On the homepage there's "model settings" where you can change the number of threads

**Comment by johncadengo** [comment]
I see, thanks @voarsh2. Let me know if you figure out a way to make it more performant on your CPUs since I have a few servers w/ similar CPUs as the ones you mentioned (Xeon E5-2600 v2 series).

**Comment by voarsh2** [comment]
> I see, thanks @voarsh2. Let me know if you figure out a way to make it more performant on your CPUs since I have a few servers w/ similar CPUs as the ones you mentioned (Xeon E5-2600 v2 series).

Haha sure, hopefully the maintainer can work it out, from this issue. I'm genuinely curious to know how the maintainer got the response speed he gets from his gif/demo...... It's not like anyone in this issue is trying to run it on a Raspberry Pi. Lol. People are using EPYC and Ryzen CPU's.......

I am going to try dalai next - maybe some better luck with a different codebase.

> I discovered that the problem is in the compilation of the new version of llama.cpp, in which the parameters passed by the compiler are making the software slower, and that an older version like: https://github.com/nomic-ai/gpt4all
> it works faster because it doesn't have these optimizations

You might help the maintainer by providing some specifics on these optimisations and your proof?

**Comment by BrunoIsaac27** [comment]
you can try to read the [thread](https://github.com/ggerganov/llama.cpp/issues/603)

**Comment by johncadengo** [comment]
There's a new PR merged into Llama.cpp: https://github.com/ggerganov/llama.cpp/pull/613

I was able to compile this on my servers and have it work quite well. Is there any way we can get the latest version of llama.cpp in Serge? Might solve all the too slow issues. Just have to make sure you use the script `migrate-ggml-2023-03-30-pr613.py` to migrate the models to work with the new file format.

More context here: https://github.com/ggerganov/llama.cpp/discussions/638#discussioncomment-5492916

**Comment by johncadengo** [comment]
Looks like the latest PR incorporated updates for the new change to llama.cpp: https://github.com/nsarrazin/serge/pull/118

I'll try it out today and let you know if it helps @voarsh2

**Comment by voarsh2** [comment]
> cc @voarsh2, one thing to note is that by default is use 4 threads. I've increased that to the max number on my machines (32 in my particular case), and I started with `GPT4All` as the model, since it's a much smaller model and more performant. Getting great results with this test.

Hmmm. GPT4ALL is MUCH faster.......... 👁️ 
Still rough around the edges. :D

**Comment by BrunoIsaac27** [comment]
> I'm pleased to report that as of the latest commit ([cf84d0c](https://github.com/nsarrazin/serge/commit/cf84d0c7f52454657457dab57cbd6325777f58cc)) the performance is much better, at least on my CPUs, which were impossibly slow before.
> 
> cc @voarsh2, one thing to note is that by default is use 4 threads. I've increased that to the max number on my machines (32 in my particular case), and I started with `GPT4All` as the model, since it's a much smaller model and more performant. Getting great results with this test.
> 
> By the way, is there a way to get the default threads to be set according to the number of threads available?

what is your hardware? how long does it take to answer you? I'm running on an amd epyc vps with 6 cores and 16gb of ram and it seems to respond after 3 minutes.

**Comment by voarsh2** [comment]
> what is your hardware? how long does it take to answer you? I'm running on an amd epyc vps with 6 cores and 16gb of ram and it seems to respond after 3 minutes.

"...giving it 32 cores of CPU (32 x Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz (2 Sockets)) and on a 48 x Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz (2 Sockets) - I have a Ryzen 5x (12 x AMD Ryzen 5 3600X 6-Core Processor (1 Socket)) system ..."

"I have an RTX 260 GPU that I can't use for this project... a shame."

**Comment by BrunoIsaac27** [comment]
> > what is your hardware? how long does it take to answer you? I'm running on an amd epyc vps with 6 cores and 16gb of ram and it seems to respond after 3 minutes.
> 
> "...giving it 32 cores of CPU (32 x Intel(R) Xeon(R) CPU E5-2650 v2 @ 2.60GHz (2 Sockets)) and on a 48 x Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz (2 Sockets) - I have a Ryzen 5x (12 x AMD Ryzen 5 3600X 6-Core Processor (1 Socket)) system ..."
> 
> "I have an RTX 260 GPU that I can't use for this project... a shame."

from what I understand, your machine is not virtualized. Right? Well, how long can the model answer this question: "could you explain to me in detail how bitcoin works? I would like a technical article in a language for laymen." Could you test it for me?

**Comment by voarsh2** [comment]
> could you explain to me in detail how bitcoin works? I would like a technical article in a language for laymen.

4 threads, using 13B model took about 6 minutes to show any text. Excluding the initial read from disk as I had sent a chat before (but it still loads/unloads lots of memory, but not reading from disk......)

About 2 minutes to print out this incomplete text:
"Bitcoins are digital currency that can be used as payment online or offline, just like cash and credit cards today. Bitcoins use peer-to-peer technology to operate with no central authority; managing transactions and the issuing of bitcoins is carried out collectively by the network.Bit"

**Comment by BrunoIsaac27** [comment]
> > could you explain to me in detail how bitcoin works? I would like a technical article in a language for laymen.
> 
> 4 threads, using 13B model took about 6 minutes to show any text. Excluding the initial read from disk as I had sent a chat before (but it still loads/unloads lots of memory, but not reading from disk......)
> 
> About 2 minutes to print out this incomplete text: "Bitcoins are digital currency that can be used as payment online or offline, just like cash and credit cards today. Bitcoins use peer-to-peer technology to operate with no central authority; managing transactions and the issuing of bitcoins is carried out collectively by the network.Bit"

it took 16 minutes here with 6 threads to generate this text: "Surely, Bitcoins work through cryptography and blockchain technology that enables them to be transferred from one user's wallet to another without any middleman or central authority involved. It is an open-source software which can run on anyone's computer hardware with a high degree of security as it uses peer-to-peer networking, making the transactions transparent and verifiable for all users in real time through distributed ledger technology (DLT)."

**Comment by gaby** [comment]
We can take this discussion to Discord
